# 4. Functions

### Objectives 
* Create custom functions.
* Import and use Python Standard Library modules, such as `random` and `math`, to reuse code and avoid “reinventing the wheel.”
* Pass data between functions.
* Generate a range of random numbers. 

### Objectives (cont.)
* Learn simulation techniques using random-number generation.
* Seed the random number generator to ensure reproducibility.
* Pack values into a tuple and unpack values from a tuple.
* Return multiple values from a function via a tuple.
* Understand how an identifier’s scope determines where in your program you can use it.

### Objectives (cont.)
* Create functions with default parameter values.
* Call functions with keyword arguments.
* Create functions that can receive any number of arguments.
* Use methods of an object.

### Outline
* [4.1   	Introduction](04_01.ipynb)
* [4.2   	Defining Functions](04_02.ipynb)
* [4.3   	Functions with Multiple Parameters](04_03.ipynb)
* [4.4   	Random-Number Generation](04_04.ipynb)
* [4.5   	Case Study: A Game of Chance](04_05.ipynb)
* [4.6   	Python Standard Library](04_06.ipynb)
* [4.7   	`math` Module Functions](04_07.ipynb)
* [4.8   	Using IPython Tab Completion for Discovery](04_08.ipynb)
* [4.9   	Default Parameter Values](04_09.ipynb)
* [4.10   	Keyword Arguments](04_10.ipynb)
* [4.11   	Arbitrary Argument Lists](04_11.ipynb)
* [4.12   	Methods: Functions That Belong to Objects](04_12.ipynb)
* [4.13   	Scope Rules](04_13.ipynb)
* [4.14   	import: A Deeper Look](04_14.ipynb)
* [4.15   	Passing Arguments to Functions: A Deeper Look](04_15.ipynb)
* [4.16   	Function-Call Stack](04_16.ipynb)
* [4.17   	Functional-Style Programming](04_17.ipynb)
* [4.18   	Intro to Data Science: Measures of Dispersion](04_18.ipynb)
* 4.19   	Wrap-Up
* Exercises


# 4.2 Defining Functions
* `square` function that calculates the square of its argument.

In [1]:
def square(number):
    """Calculate the square of number."""
    return number ** 2

In [2]:
square(7)

49

In [3]:
square(2.5)

6.25

* Calling `square` with a non-numeric argument like `'hello'` causes a `TypeError` because the exponentiation operator (`**`) works only with numeric values

### Defining a Custom Function
* Definition begins with the (**`def` keyword**, followed by the function name, a set of parentheses and a colon (`:`). 
* By convention function names should begin with a lowercase letter and in multiword names underscores should separate each word. 
* Required parentheses contain the function’s **parameter list**.
* Empty parentheses mean no parameters. 
* The indented lines after the colon (`:`) are the function’s **block**
    * Consists of an optional docstring followed by the statements that perform the function’s task.

### Specifying a Custom Function’s Docstring 
* _Style Guide for Python Code_: First line in a function’s block should be a docstring that briefly explains the function’s purpose.

### Returning a Result to a Function’s Caller
* Function calls also can be embedded in expressions:

In [4]:
print('The square of 7 is', square(7))

The square of 7 is 49


* Three Ways to Return a Result to a Function’s Caller
    * **`return`** followed by an expression.
    * **`return`** without an expression implicitly returns **`None`**&mdash;represents the **absence of a value** and **evaluates to `False` in conditions**.
    * **No `return` statement implicitly returns `None`**.


### What Happens When You Call a Function
* Parameters exist only during the function call. 
* Created on each call to the function to receive arguments.
* Destroyed when the function returns its result to the caller. 
* A function’s parameters and variables defined in its block are all **local variables**.

### Accessing a Function’s Docstring Via IPython’s Help Mechanism 
* Following a function's name with `?` in IPython displays its docstring:

In [5]:
square?

Signature: square(number)
Docstring: Calculate the square of number.
File:      ~/Dropbox/books/2019/Python/PyCDS_JupyterSlides/ch04/<ipython-input-1-7d5dc51751d0>
Type:      function


* If the function’s source code is accessible from IPython, `??` displays the function’s docstring and full source-code definition:

In [6]:
square??

Signature: square(number)
Source:   
def square(number):
    """Calculate the square of number."""
    return number ** 2
File:      ~/Dropbox/books/2019/Python/PyCDS_JupyterSlides/ch04/<ipython-input-1-7d5dc51751d0>
Type:      function


# 4.3 Functions with Multiple Parameters
* `maximum` function that determines and returns the largest of three values.

In [1]:
def maximum(value1, value2, value3):
    """Return the maximum of three values."""
    max_value = value1
    if value2 > max_value:
        max_value = value2
    if value3 > max_value:
        max_value = value3
    return max_value

In [2]:
maximum(12, 27, 36)

36

In [3]:
maximum(12.3, 45.6, 9.7)

45.6

In [4]:
maximum('yellow', 'red', 'orange')

'yellow'

* We did not place blank lines above and below the `if` statements, because pressing return on a blank line in interactive mode completes the function’s definition.
* May call maximum with mixed types, such as `int`s and `float`s.

In [5]:
maximum(13.5, -3, 7)

13.5

### Function maximum’s Definition
* Specify multiple parameters in a comma-separated list.
* To determine the largest value:
    * Assume that `value1` contains the largest value. 
    * The first `if` statement then tests `value2 > max_value`, and if this condition is `True` assigns `value2` to `max_value`. 
    * The second `if` statement then tests `value3 > max_value`, and if this condition is `True` assigns `value3` to `max_value`. 
* Now, `max_value` contains the largest value.

### Python’s Built-In max and min Functions
* For many common tasks, the capabilities you need already exist in Python. 
* Built-in `max` and `min` functions know how to determine the largest and smallest of their two or more arguments:

In [6]:
max('yellow', 'red', 'orange', 'blue', 'green')

'yellow'

In [7]:
min(15, 9, 27, 14)

9

* Each also can receive an iterable argument, such as a list or a string. 
* Using built-in functions or functions from the Python Standard Library’s modules rather than writing your own can reduce development time and increase program reliability, portability and performance. 
* [Python’s built-in functions and modules](https://docs.python.org/3/library/index.html)

# 4.4 Random-Number Generation
* Can introduce the element of chance via the Python Standard Library’s `random` module. 

### Rolling a Six-Sided Die
* Product 10 random integers in the range 1–6 to simulate rolling a six-sided die:

In [1]:
import random

In [2]:
for roll in range(10):
    print(random.randrange(1, 7), end=' ')

2 4 1 5 1 4 3 6 4 5 

* `randrange` function generates an integer from the first argument value up to, but not including, the second argument value.
* Different values are displayed if you re-execute the loop.

In [3]:
for roll in range(10):
    print(random.randrange(1, 7), end=' ')

6 5 5 1 3 4 1 5 5 2 

* Can guarantee **reproducibility** of a random sequence with the `random` module’s `seed` function.  

### Rolling a Six-Sided Die 6,000,000 Times
* If `randrange` truly produces integers at random, every number in its range has an equal probability (or chance or likelihood) of being returned each time we call it. 
* Roll a die 6,000,000 times.
* Each die face should occur approximately 1,000,000 times.
* We used Python’s underscore (_) digit separator to make the value 6000000 more readable. 

```python
### Rolling a Six-Sided Die 6,000,000 Times

# fig04_01.py
"""Roll a six-sided die 6,000,000 times."""
import random

# face frequency counters
frequency1 = 0
frequency2 = 0
frequency3 = 0
frequency4 = 0
frequency5 = 0
frequency6 = 0

# 6,000,000 die rolls
for roll in range(6_000_000):  # note underscore separators
    face = random.randrange(1, 7)

    # increment appropriate face counter
    if face == 1:
        frequency1 += 1
    elif face == 2:
        frequency2 += 1
    elif face == 3:
        frequency3 += 1
    elif face == 4:
        frequency4 += 1
    elif face == 5:
        frequency5 += 1
    elif face == 6:
        frequency6 += 1

print(f'Face{"Frequency":>13}')
print(f'{1:>4}{frequency1:>13}')
print(f'{2:>4}{frequency2:>13}')
print(f'{3:>4}{frequency3:>13}')
print(f'{4:>4}{frequency4:>13}')
print(f'{5:>4}{frequency5:>13}')
print(f'{6:>4}{frequency6:>13}')
```

In [4]:
run fig04_01.py

Face    Frequency
   1       999952
   2       999497
   3      1001044
   4      1000356
   5      1000563
   6       998588


### Seeding the Random-Number Generator for Reproducibility
* Function `randrange` generates pseudorandom numbers. 
* Numbers appear to be random, because each time you start a new interactive session or execute a script that uses the random module’s functions, Python internally uses a different seed value. 
* When you’re debugging logic errors in programs that use randomly generated data, it can be helpful to use the same sequence of random numbers. 
* To do this, use the random module’s `seed` function to seed the random-number generator:

In [5]:
random.seed(32)

In [6]:
for roll in range(10):
    print(random.randrange(1, 7), end=' ')

1 2 2 3 6 2 4 1 6 1 

In [7]:
for roll in range(10):
    print(random.randrange(1, 7), end=' ')

1 3 5 3 1 5 6 4 3 5 

In [8]:
random.seed(32)

In [9]:
for roll in range(10):
    print(random.randrange(1, 7), end=' ')

1 2 2 3 6 2 4 1 6 1 

# 4.5 Case Study: A Game of Chance
* Requirements statement:
>You roll two six-sided dice, each with faces containing one, two, three, four, five and six spots, respectively. When the dice come to rest, the sum of the spots on the two upward faces is calculated. If the sum is 7 or 11 on the first roll, you win. If the sum is 2, 3 or 12 on the first roll (called “craps”), you lose (i.e., the “house” wins). If the sum is 4, 5, 6, 8, 9 or 10 on the first roll, that sum becomes your “point.” To win, you must continue rolling the dice until you “make your point” (i.e., roll that same point value). You lose by rolling a 7 before making your point.
* Script that simulates the game:

```python
# fig04_02.py
"""Simulating the dice game Craps."""
import random

def roll_dice():
    """Roll two dice and return their face values as a tuple."""
    die1 = random.randrange(1, 7)
    die2 = random.randrange(1, 7)
    return (die1, die2)  # pack die face values into a tuple

def display_dice(dice):
    """Display one roll of the two dice."""
    die1, die2 = dice  # unpack the tuple into variables die1 and die2
    print(f'Player rolled {die1} + {die2} = {sum(dice)}')

die_values = roll_dice()  # first roll
display_dice(die_values)

# determine game status and point, based on first roll
sum_of_dice = sum(die_values)

if sum_of_dice in (7, 11):  # win
    game_status = 'WON'
elif sum_of_dice in (2, 3, 12):  # lose
    game_status = 'LOST'
else:  # remember point
    game_status = 'CONTINUE'
    my_point = sum_of_dice
    print('Point is', my_point)

# continue rolling until player wins or loses
while game_status == 'CONTINUE':
    die_values = roll_dice()
    display_dice(die_values)
    sum_of_dice = sum(die_values)

    if sum_of_dice == my_point:  # win by making point
        game_status = 'WON'
    elif sum_of_dice == 7:  # lose by rolling 7
        game_status = 'LOST'

# display "wins" or "loses" message
if game_status == 'WON':
    print('Player wins')
else:
    print('Player loses')
```

In [1]:
run fig04_02.py

Player rolled 5 + 4 = 9
Point is 9
Player rolled 6 + 3 = 9
Player wins


In [2]:
run fig04_02.py

Player rolled 6 + 6 = 12
Player loses


In [3]:
run fig04_02.py

Player rolled 4 + 3 = 7
Player wins


In [5]:
run fig04_02.py

Player rolled 1 + 5 = 6
Point is 6
Player rolled 5 + 4 = 9
Player rolled 6 + 3 = 9
Player rolled 1 + 6 = 7
Player loses


### Function `roll_dice`—Returning Multiple Values Via a Tuple
* Simulates rolling two dice on each roll. 
* Sometimes it’s useful to return more than one value, as in `roll_dice`, which returns both die values as a **tuple**—an immutable sequences of values. 
* To create a tuple, separate its values with commas&mdash;known as **packing a tuple**. 
* Parentheses are optional, but we recommend using them for clarity. 

### Function `display_dice`
* Assigning a tuple to a comma-separated list of variables **unpacks the tuple**. 
* Number of variables to the left of `=` must match the number of elements in the tuple; otherwise, a `ValueError occurs`.

### First Roll
* When the script begins executing, we roll the dice and display the results. 
* You can win or lose on the first roll or any subsequent roll&mdash;`game_status` keeps track of the win/loss status.
* The `in` operator in the following expression tests whether the tuple `(7, 11)` contains `sum_of_dice`’s value
```python
sum_of_dice in (7, 11)
```
* The operator’s right operand can be any iterable. 
* The preceding concise condition is equivalent to
```python
(sum_of_dice == 7) or (sum_of_dice == 11) 
```

### Subsequent Rolls 


### Displaying the Final Results 


# 4.7 `math` Module Functions
* The **`math` module** defines functions for performing various common mathematical calculations. 

In [ ]:
import math

In [ ]:
math.sqrt(900)

In [ ]:
math.fabs(-10)

* Some `math` module functions are summarized below
* [View the complete list](https://docs.python.org/3/library/math.html)

Function&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;| Description | Example &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
:------- | :------- | :--------
`ceil(`x`)`	| Rounds x to the smallest integer not less than x	| `ceil(9.2) is 10.0`, `ceil(-9.8) is -9.0`
`floor(`x`)`	| Rounds x to the largest integer not greater than x	| `floor(9.2) is 9.0`, `floor(-9.8) is -10.0`
`sin(`x`)`	| Trigonometric sine of x (x in radians)	| `sin(0.0) is 0.0`
`cos(`x`)`	| Trigonometric cosine of x (x in radians)	| `cos(0.0) is 1.0`
`tan(`x`)` 	| Trigonometric tangent of x (x in radians)	| `tan(0.0) is 0.0`
`exp(`x`)`	| Exponential function e<sup>x</sup>	| `exp(1.0) is 2.718282`, `exp(2.0) is 7.389056`
`log(`x`)`	| Natural logarithm of x (base e)	| `log(2.718282) is 1.0`, `log(7.389056) is 2.0`
`log10(`x`)`	| Logarithm of x (base 10)	| `log10(10.0) is 1.0`, `log10(100.0) is 2.0`
`pow(`x, y`)`	| x raised to power y (x<sup>y</sup>)	| `pow(2.0, 7.0) is 128.0`, `pow(9.0, .5) is 3.0`
`sqrt(`x`)`	| square root of x	| `sqrt(900.0) is 30.0`, `sqrt(9.0) is 3.0`
`fabs(`x`)`	| Absolute value of x—always returns a float. Python also has the built-in function `abs`, which returns an `int` or a `float`, based on its argument.	| `fabs(5.1) is 5.1`, `fabs(-5.1) is 5.1` 
`fmod(`x, y`)`	| Remainder of x/y as a floating-point number	| `fmod(9.8, 4.0) is 1.8`

# 4.9 Default Parameter Values
* You can specify that a parameter has a **default parameter value**. 
* When calling the function, if you omit the argument for a parameter with a default parameter value, the default value for that parameter is automatically passed. 

In [ ]:
def rectangle_area(length=2, width=3):
    """Return a rectangle's area."""
    return length * width

* Specify a default parameter value by following a parameter’s name with an `=` and a value. 
* Any parameters with default parameter values must appear in the parameter list to the _right_ of parameters that do not have defaults. 

In [ ]:
rectangle_area()

In [ ]:
rectangle_area(10)

In [ ]:
rectangle_area(10, 5)

# 4.10 Keyword Arguments
* When calling functions, you can use **keyword arguments** to pass arguments in `any` order.

In [ ]:
def rectangle_area(length, width):
    """Return a rectangle's area."""
    return length * width

* Each keyword _argument in a call_ has the form _parametername=value_. 
* Order of keyword arguments does not matter.

In [ ]:
rectangle_area(width=5, length=10)

# 4.11 Arbitrary Argument Lists
* Functions with **arbitrary argument lists**, such as built-in functions `min` and `max`, can receive _any_ number of arguments. 
* Function `min`'s documentation states that `min` has two _required_ parameters (named `arg1` and `arg2`) and an optional third parameter of the form **`*args`**, indicating that the function can receive any number of additional arguments. 
* The `*` before the parameter name tells Python to pack any remaining arguments into a tuple that’s passed to the `args` parameter.

### Defining a Function with an Arbitrary Argument List
* `average` function that can receive any number of arguments.

In [ ]:
def average(*args):
    return sum(args) / len(args)

* The `*args` parameter must be the _rightmost_ parameter. 

In [ ]:
average(5, 10)

In [ ]:
average(5, 10, 15)

In [ ]:
average(5, 10, 15, 20)

### Passing an Iterable’s Individual Elements as Function Arguments 
* Can unpack a tuple’s, list’s or other iterable’s elements to pass them as individual function arguments. 
* The **`*` operator**, when applied to an iterable argument in a function call, unpacks its elements. 

In [ ]:
grades = [88, 75, 96, 55, 83]

In [ ]:
average(*grades)

* Equivalent to `average(88, 75, 96, 55, 83)`.

# 4.12 Methods: Functions That Belong to Objects
* A **method** is simply a function that you call on an object using the form 
```
object_name.method_name(arguments)
```

In [ ]:
s = 'Hello'

In [ ]:
s.lower()  # call lower method on string object s

In [ ]:
s.upper()

In [ ]:
s

# 4.13 Scope Rules
* Each identifier has a `scope` that determines where you can use it in your program.
* [Complete scope details](https://docs.python.org/3/tutorial/classes.html#python-scopes-and-namespaces).

### Local Scope
* A local variable’s identifier has **local scope**. 

### Global Scope
* Identifiers defined outside any function (or class) have **global scope**—these may include functions, variables and classes.

### Accessing a Global Variable from a Function 

In [ ]:
x = 7

In [ ]:
def access_global():
    print('x printed from access_global:', x)

In [ ]:
access_global()

* By default, you cannot _modify_ a global variable in a function
* Python creates a **new local variable** when you first assign a value to a variable in a function’s block.
* In function `try_to_modify_global`’s block, the local `x` **shadows** the global `x`, making it inaccessible in the scope of the function’s block. 

In [ ]:
def try_to_modify_global():
    x = 3.5
    print('x printed from try_to_modify_global:', x)

In [ ]:
try_to_modify_global()

In [ ]:
x

* To modify a global variable in a function’s block, you must use a **`global`** statement to declare that the variable is defined in the global scope:

In [ ]:
def modify_global():
    global x;
    x = 'hello'
    print('x printed from modify_global:', x)
    

In [ ]:
modify_global()

In [ ]:
x

### Blocks vs. Suites 
* When you create a variable in a block, it’s _local_ to that block.
* When you create a variable in a control statement’s suite, the variable’s scope depends on where the control statement is defined:
    * If it's in the global scope, any variables defined in the control statement have **global scope**.
    * If it's in a function’s block, any variables defined in the control statement have **local scope**.

### Shadowing Functions
* In the preceding chapters, when summing values, we stored the sum in a variable named `total`. 
* If you define a variable named `sum`, it _shadows_ the built-in function `sum`, making it inaccessible in your code. 

In [ ]:
sum = 10 + 5

In [ ]:
sum

In [ ]:
sum([10, 5])

### Statements at Global Scope 
* Script statements at global scope execute as soon as they’re encountered by the interpreter, whereas statements in a block execute only when the function is called. 

# 4.14 `import`: A Deeper Look

### Importing Multiple Identifiers from a Module
* Use `from…import` to import a comma-separated list of identifiers from a module then use them without having to precede them with the module name and a dot (`.`):

In [ ]:
from math import ceil, floor

In [ ]:
ceil(10.3)

In [ ]:
floor(10.7)

### Caution: Avoid Wildcard Imports 
* Import _all_ identifiers defined in a module with a **wildcard** **`import`**.
* Makes all of the module’s identifiers available. 
* Can lead to subtle errors.
* Considered a dangerous practice. 

In [ ]:
e = 'hello'

In [ ]:
from math import *

In [ ]:
e

* After executing the import, variable `e` is replaced, possibly by accident, with the `math` module’s constant `e`.

### Binding Names for Modules and Module Identifiers
* Sometimes it’s helpful to import a module and use an abbreviation for it to simplify your code. 
* The `import` statement’s **`as`** clause allows you to specify the name used to reference the module’s identifiers. 


In [ ]:
import statistics as stats

In [ ]:
grades = [85, 93, 45, 87, 93]

In [ ]:
stats.mean(grades)

# 4.15 Passing Arguments to Functions: A Deeper Look 
* **Python arguments are always passed by reference**. 
* Some people call this **pass-by-object-reference**, because “everything in Python is an object.” 
* When a function call provides an argument, Python copies the argument object’s _reference_—not the object itself—into the corresponding parameter.

### Memory Addresses, References and “Pointers”
* After an assignment like the following, the variable `x` contains a reference to an _object_ containing `7` stored _elsewhere_ in memory.

![Variable referring to an object](ch04images/AAEMYQU0a.png "Variable referring to an object")

### Built-In Function id and Object Identities 
* Every object has a **unique** **identity**&mdash;an `int` value which **identifies only that object** while it remains in memory.
* **Built-in `id` function** to obtain an object's identity.


In [1]:
x = 7

In [2]:
id(x)

140704917108704

### Passing an Object to a Function 

In [3]:
def cube(number):
    print('id(number):', id(number))
    return number ** 3

In [4]:
cube(x)

id(number): 140704917108704


343

* The identity displayed for `cube`’s parameter `number` is the _same_ as that displayed for `x` previously. 
* The _argument_ `x` and the _parameter_ `number` refer to the _same object_ while `cube` executes. 

### Testing Object Identities with the is Operator 
* The **`is`** **operator** returns `True` if its two operands have the _same identity_:

In [5]:
def cube(number):
    print('number is x:', number is x)  # x is a global variable
    return number ** 3

In [6]:
cube(x)

number is x: True


343

### Immutable Objects as Arguments
* When a function receives as an argument a reference to an _immutable_ (unmodifiable) object, even though you have direct access to the original object in the caller, you cannot modify the original immutable object’s value.
* Note: when you pass a global var to a function, the function is not able to modify that variable, therefore that variaible is immutable. What looks like a modification, is actually a local change for new variable.

In [7]:
def cube(number):
    print('id(number) before modifying number:', id(number))
    number **= 3
    print('id(number) after modifying number:', id(number))
    return number

In [8]:
cube(x)

id(number) before modifying number: 140704917108704
id(number) after modifying number: 1730314759440


343

In [9]:
print(f'x = {x}; id(x) = {id(x)}')

x = 7; id(x) = 140704917108704


### Mutable Objects as Arguments
* We’ll show that when a reference to a _mutable_ object like a list is passed to a function, the function _can_ modify the original object in the caller. 

# 4.16 Function-Call Stack
* A stack is like a pile of dishes. 
* When you add a dish to the pile, you place it on the top. 
* When you remove a dish from the pile, you take it from the top. 
* Stacks are last-in, first-out (LIFO) data structures.

### Stacks and Your Web Browser’s Back Button
* A stack of webpage addresses supports a browser’s back button. 
* For each new web page you visit, the browser pushes the address of the page you were viewing onto the back button’s stack. 
* When you press the browser’s back button, the browser pops the top stack element to get the prior web page’s address, then displays that web page.

### Stack Frames
* Function-call stack supports the function call/return mechanism. 
* Eventually, each function must return program control to the point at which it was called. 
* For each function call, the interpreter pushes an entry called a stack frame (or an activation record) onto the stack. 
* Contains the return location that the called function needs so it can return control to its caller. 
* When the function finishes executing, the interpreter pops the function’s stack frame, and control transfers to the return location that was popped.
* The top stack frame always contains the information the currently executing function needs to return control to its caller. 

### Local Variables and Stack Frames
* Most functions have one or more parameters and possibly local variables that need to:
    * exist while the function is executing,
    * remain active if the function makes calls to other functions, and
    * “go away” when the function returns to its caller.
* A stack frame stores the function’s local variables. 

### Stack Overflow
* The amount of memory in a computer is finite, so only a certain amount of memory can be used to store stack frames on the function-call stack. 
* If the function-call stack runs out of memory as a result of too many function calls, a fatal error known as stack overflow occurs.

### Principle of Least Privilege
* States that code should be granted only the amount of privilege and access that it needs to accomplish its designated task, but no more. 
* An example of this is the scope of a local variable, which should not be visible when it’s not needed. 
* This is why a function’s local variables are placed in stack frames on the function-call stack, so they can be used by that function while it executes and go away when it returns. 
* The principle of least privilege  your programs more robust by preventing code from accidentally (or maliciously) modifying variable values that should not be accessible to it.

# 4.17 Functional-Style Programming
* Python offers “functional-style” features that help you write code which is less likely to contain errors, more concise and easier to read, debug and modify. 
* The chart below lists most of Python’s key functional-style programming capabilities and shows in parentheses the chapters in which we initially cover many of them.

| Functional-style programming topics
| -----------------------
| avoiding side effects (4)
| closures 
| declarative programming (4)
| decorators (10)
| dictionary comprehensions (6)
| `filter`/`map`/`reduce` (5)
| `functools` module 
| generator expressions (5)
| generator functions
| higher-order functions (5)
| immutability (4)
| internal iteration (4)
| iterators (3)
| `itertools` module (16)
| `lambda` expressions (5)
| lazy evaluation (5)
| list comprehensions (5)
| operator module (5, 11, 16)
| pure functions (4)
| range function (3, 4)
| reductions (3, 5)
| set comprehensions (6)

### What vs. How
* Functional-style programming lets you simply say _what_ you want to do. 
* Typically, library code handles the _how_ for you. 
* Can eliminate many errors. 
* Functional-style programming emphasizes **immutability**. 
* Python’s `for` statement and `range` function _hide_ most counter-controlled iteration details. 
    * You specify _what_ values `range` should produce and the variable that should receive each value as it’s produced. 
    * Function `range` _knows how_ to produce those values. 
    * The `for` statement _knows how_ to get each value from `range` and _how_ to stop iterating when there are no more values. 
* Specifying _what_, but not _how_, is an important aspect of **internal iteration**—a key functional-style programming concept. 
* Stating what you _want_ done rather than programming _how_ to do it is known as **declarative programming**. 

### Pure Functions
* A **pure function’s** result depends only on the argument(s) you pass to it. 
* Given a particular argument (or arguments), a pure function always produces the same result. 
* A pure function does not have _side effects_. 

In [ ]:
values = [1, 2, 3]

In [ ]:
sum(values)

In [ ]:
sum(values)  # same call always returns same result

In [ ]:
values

# 4.18 Intro to Data Science: Measures of Dispersion
* Considered the measures of central tendency—mean, median and mode. 
* Help us categorize typical values in a group.
* An entire group is called a **population**. 
* Sometimes a population is quite large, such as the people likely to vote in the next U.S. presidential election, which is a number in excess of 100,000,000 people. 
* For practical reasons, the polling organizations trying to predict who will become the next president work with carefully selected small subsets of the population known as **samples**. 
* Hear we introduce **measures of dispersio**n (also called **measures of variability**) that help you understand how **spread out** the values are. 
* We’ll calculate each measure of dispersion both by hand and with functions from the module `statistics`, using the following population of 10 six-sided die rolls:
> 1, 3, 4, 2, 6, 5, 3, 4, 5, 2

### Variance
* To determine variance, begin with the mean of these values—3.5. 
* Next, subtract the mean from every die value:
> -2.5, -0.5, 0.5, -1.5, 2.5, 1.5, -0.5, 0.5, 1.5, -1.5
* Then, square each of these results (yielding only positives):
> 6.25, 0.25, 0.25, 2.25, 6.25, 2.25, 0.25, 0.25, 2.25, 2.25
* Finally, calculate the mean of these squares, which is 2.25 (22.5 / 10)—this is the **population
variance**. 
* Squaring the difference between each die value and the mean of all die values emphasizes **outliers**—the values that are farthest from the mean—which can be important in data analysis.
* The following code uses the `statistics` module’s `pvariance` function to confirm our manual result:

In [ ]:
import statistics

In [ ]:
statistics.pvariance([1, 3, 4, 2, 6, 5, 3, 4, 5, 2])

### Standard Deviation
* The standard deviation is the square root of the variance (in this case, 1.5), which tones
down the effect of the outliers. 
* The smaller the variance and standard deviation are, the closer the data values are to the mean and the less overall dispersion (that is, spread) there is between the values and the mean. 
* The following code calculates the population standard deviation with the `statistics` module’s `pstdev` function, confirming our manual
result:

In [ ]:
statistics.pstdev([1, 3, 4, 2, 6, 5, 3, 4, 5, 2])

In [ ]:
import math

In [ ]:
math.sqrt(statistics.pvariance([1, 3, 4, 2, 6, 5, 3, 4, 5, 2]))


### Advantage of Population Standard Deviation vs. Population Variance
* Suppose you’ve recorded the March Fahrenheit temperatures in your area. 
* You might have 31 numbers such as 19, 32, 28 and 35. 
* The units for these numbers are degrees.
* When you square your temperatures to calculate the population variance, the units of the population variance become **“degrees squared.”**
* When you take the square root of the population variance to calculate the population standard deviation, the units once again become **degrees**, which are the same units as your temperatures.